apps/emqx_rule_engine/include/rule_engine.hrl

```diff
@@ -200,3 +198,61 @@
 -define(RES_PARAMS_TAB, emqx_resource_params).
 -define(RULE_HOOKS, emqx_rule_hooks).
 -define(RES_TYPE_TAB, emqx_resource_type).
+
+%% execute Continuation if Operator succeeds
+-define(EITHER(Operator, Continuation),
+        case Operator of
+            ok ->
+                Continuation;
+            {error, _} = _ERError_ ->
+                _ERError_
+        end).
+
+%% if Operator succeeds, bind its result to Bind, and execute Continuation
+-define(E_BIND(Bind, Operator, Continuation),
+        case Operator of
+            {ok, Bind} ->
+                Continuation;
+            {error, _} = _EBError_ ->
+                _EBError_
+        end).
+
+-define(E_LET(Bind, Operator, Continuation),
+        case Operator of
+            {ok, Bind} ->
+                Continuation;
+            _ELError_ ->
+                {error, _ELError_}
+        end).
+
+%% safely try for sensitive data
+-define(SAFELY_TRY(Try, Catch),
+        try
+            Try
+        catch _Error_:_Reason_:_UnSafeTrace_ ->
+                _Trace_ = emqx_secret:prune_stacktrace(_UnSafeTrace_),
```

Comment:
could you provide an example in which stacktrace exposes the function arguments.?

Reply:
<img width="1394" alt="image" src="https://user-images.githubusercontent.com/22971499/196315806-b4952f7d-a05e-425e-9a4e-1f6cc93c14b8.png">

Reply:
<img width="1395" alt="image" src="https://user-images.githubusercontent.com/22971499/196316510-39db8009-a035-43f8-9564-781ad8c7473d.png">
Looks like I'm doing something wrong...